lib/client.ml

```diff
@@ -95,6 +130,171 @@ let route client router msg =
+        recipient = recipient.address;
+        payload = Util.Encoding.pack bin_writer_message message;
+      }
+
+  let send_message message client (recipient : Peer.t) =
+    let message = create_message client message recipient in
+    send_to client message
+
+  let send_ping_to client peer = send_message Ping client peer
+
+  let send_acknowledge_to client peer = send_message Acknowledge client peer
+
+  let send_ping_request_to client (recipient : Peer.t) =
+    send_message (PingRequest recipient.address) client recipient
+
+  (** Processes an incoming message bound for the failure detector of a client *)
+  let handle_message client message =
+    let open Message in
+    let peer_of_client = peer_from !client in
+    let peer = Peer.from message.sender in
+    let msg = Util.Encoding.unpack bin_read_message message.payload in
+    let t = !client.failure_detector in
+    match msg with
+    | Ping -> send_acknowledge_to client peer
+    | PingRequest addr -> (
+      let new_seq_no = next_seq_no t in
+      let peer_opt = Peer.get_neighbor peer_of_client addr in
+      match peer_opt with
+      | Some peer -> send_ping_request_to client peer
+      | None -> (
+        match%lwt wait_ack_timeout t new_seq_no t.config.protocol_period with
+        | Ok _ -> Lwt.return ()
+        | Error _ -> send_acknowledge_to client peer))
+    | Acknowledge ->
+    match Base.Hashtbl.find t.acknowledges t.sequence_number with
+    | Some cond ->
+      Lwt_condition.broadcast cond ();
+      Lwt.return ()
+    | None -> Lwt.return ()
+
+  (** This function will be called by failure_detection 
+  at each round of the protocol, and update the peers *)
+  let probe_peer t client peer_to_update =
+    let peer_of_client = peer_from !client in
+    let new_seq_no = next_seq_no t in
+    let _ = send_ping_to client peer_to_update in
+    match%lwt wait_ack_timeout t new_seq_no t.config.round_trip_time with
+    | Ok _ -> Lwt.return ()
+    | Error _ -> (
+      let pingers =
+        t.config.peers_to_ping
+        |> pick_random_neighbors peer_of_client.neighbors
+        |> List.map Peer.from in
+      let _ = List.map (send_ping_request_to client) pingers in
+      let wait_time = t.config.protocol_period - t.config.round_trip_time in
+      match%lwt wait_ack_timeout t new_seq_no wait_time with
+      | Ok _ -> Lwt.return ()
+      | Error _ ->
+        (* TODO: Implement Suspicion Mechanism.
+           This is correct in the basic SWIM protocol, but it is a very heavy penalty.
+           When there is no ACK (direct or indirect) the peer must be set to `Suspicious`.
+           See section 4.2 from https://www.cs.cornell.edu/projects/Quicksilver/public_pdfs/SWIM.pdf *)
+        let _ = update_neighbor_status peer_of_client peer_to_update Faulty in
+        Lwt.return ())
+
+  (** High level function, which must be run within an async thread, like:
+  Lwt.async (fun () -> failure_detection t client); *)
+  let failure_detection client =
```

Comment:
I don't remember why I made it `rec` at first... But looks like this was useless. Thanks!

Reply:
Because you intended for it to run independently as though it was its own server/message handler. Now that it runs in the server (which is already recursive), it doesn't need to be recursive itself.